Decision transformer gym

edbeeching opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Added the Decision Transformer Model to the transformers library (https://github.com/kzl/decision-transformer)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->
@LysandreJik

edbeeching committed:
Created the Decision Transformer Modle

edbeeching committed:
Created the Decision Transformer Modle

edbeeching committed:
updating tests, copy to other machine

edbeeching committed:
updating tests, copy to other machine

edbeeching committed:
Added last hidden size to Decision Transformer modelling outputs

edbeeching committed:
Added last hidden size to Decision Transformer modelling outputs

edbeeching committed:
Removed copy of original DT file

edbeeching committed:
Removed copy of original DT file

edbeeching committed:
made a temporary change to gpt2 to have it conform with the Decision Transformer version

edbeeching committed:
made a temporary change to gpt2 to have it conform with the Decision Transformer version

edbeeching committed:
Updated tests

edbeeching committed:
Updated tests

edbeeching committed:
Ignoring a file used to test the DT model

edbeeching committed:
Ignoring a file used to test the DT model

edbeeching committed:
added comments to config file

edbeeching committed:
added comments to config file

edbeeching committed:
added comments and argument descriptions to decision transformer file

edbeeching committed:
added comments and argument descriptions to decision transformer file

edbeeching committed:
Updated doc

edbeeching committed:
Updated doc

edbeeching committed:
Ran "make style"

edbeeching committed:
Ran "make style"

edbeeching committed:
Remove old model imports

edbeeching committed:
Remove old model imports

edbeeching committed:
Removed unused imports, cleaned up init file

edbeeching committed:
Removed unused imports, cleaned up init file

LysandreJik commented on tests/test_modeling_decision_transformer.py:
```diff
@@ -0,0 +1,163 @@
+        return config, inputs_dict
+
+
+@require_torch
+class DecisionTransformerModelTest(  # ModelTesterMixin is removed as this model as the input / outputs of this model do not conform to a typical NLP model.
+    GenerationTesterMixin, unittest.TestCase
+):
```
Is there any additional test that would make it more robust? I think a test for values, i.e. an integration test would be nice (see other files like `tests/bert/test_modeling_bert.py`), and logic tests regarding save/load. I think you identified quite a few failing tests when adding the `ModelTesterMixin`, but I believe quite a few tests must have been passing as well. Porting these here would be ideal!

LysandreJik reviewed (commented):
This looks good! I don't have much to add, except that I think the PR could be greatly simplified by:
- It seems like most of the changes to GPT-2 could simply be achieved by having the positional word embeddings be full of zeros so that the addition does nothing.
- A lot of changes are linked to your `black` probably not seeing the line character limit is 119. You should run `make fixup` from the root of your clone after installing the dependencies with `pip install -e .[quality]`
- There's a `datasets` submodule that was added and that is not necessary 

Finally, I'm not sure the default example applies here; if possible a full working example would go a long way towards understanding how the model behaves. If you think this is better suited in a notebook, this is definitely fine too.

Pinging @patrickvonplaten and @sgugger for knowledge; no need to do a review just yet, but if you have any other ideas regarding the first RL model in the library, feel free to mention them!

edbeeching committed:
Update docs/source/model_doc/decision_transformer.mdx

added my username

Co-authored-by: Lysandre Debut <lysandre@huggingface.co>

edbeeching committed:
Update docs/source/model_doc/decision_transformer.mdx

added my username

Co-authored-by: Lysandre Debut <lysandre@huggingface.co>

edbeeching committed:
Reverted changes made to gpt2

edbeeching committed:
Reverted changes made to gpt2

edbeeching committed:
Removed datasets submodule

edbeeching committed:
Removed datasets submodule

edbeeching committed:
Update the modeling outputs to include gpt2 attentions, hidden states and last hidden states

edbeeching committed:
Update the modeling outputs to include gpt2 attentions, hidden states and last hidden states

edbeeching committed:
Added support for return of hidden states, attentions and return dict of gpt2 model.

edbeeching committed:
Added support for return of hidden states, attentions and return dict of gpt2 model.

edbeeching committed:
Updated tests to include many of the ModelTesterMixin tests. 

The following tests are skipped: test_generate_without_input_ids, test_pruning, test_resize_embeddings, test_head_masking, test_attention_outputs, test_hidden_states_output, test_inputs_embeds, test_model_common_attributes

edbeeching committed:
Updated tests to include many of the ModelTesterMixin tests. 

The following tests are skipped: test_generate_without_input_ids, test_pruning, test_resize_embeddings, test_head_masking, test_attention_outputs, test_hidden_states_output, test_inputs_embeds, test_model_common_attributes

edbeeching committed:
Added missing line to the end of gpt2 file

edbeeching committed:
Added missing line to the end of gpt2 file

edbeeching commented:
I have:
- Removed the (temporary) changes that were made on the gpt2 model
- Added zeroed position encodings to the call to gpt2 forward, so they are ignored. (as these are implemented in the DT model)
- Fixed issues related to black line length (my IDE was set to the default I think)
- Removed datasets submodule
- Added and fixed many tests for the ModelTesterMixin. I skip:
  - test_generate_without_input_ids
  - test_pruning
  - test_resize_embeddings
  - test_head_masking
  - test_attention_outputs
  - test_hidden_states_output
  - test_inputs_embeds
  - test_model_common_attributes
- Updated some of the doc strings etc.

Still Todo:

- Example file / notebook. Ideally this would show a model checkpoint being loaded and then demonstrate the performance on a RL environment. Is this what you are thinking? I have checkpoints saved from the authors implementation, but I will need to modify them so the state_dict matches this implementation (it will be missing the gpt2 position encoding weights, for example)
- Integration test. Similar issue as the above.

Please let me know if you have any comments, questions or ideas related to the commits I have just added.

patrickvonplaten commented:
Really cool to see the first RL model in Transformers here! Agree with @LysandreJik in that we can just set the position ids to a 0 valued tensor so that the model doesn't need to be adapted. 

For me, it would be important to understand how the model would be used in practice so that we can better understand how to best implement the model in Transformers. @edbeeching, do you think it could be possible to add an integration test that verifies that a **minimal** use case of the model works as expected? E.g. maybe we could show how the user would use this implementation to retrieve an action state `a_t` from `s_{t-1}`, `r_{t-1}` and `a_{t-1}` and then how `a_t` would be further treated for the next time step? So a code-snippet showing a two-step inference with some comments/code explaining where the reward and state come from as well. 

Very exciting!

edbeeching committed:
Added an integration test for the Decision Transformer

Test performs and autoregressive evaluation for two time steps

edbeeching committed:
Added an integration test for the Decision Transformer

Test performs and autoregressive evaluation for two time steps

edbeeching committed:
Set done and info to _ to fix failing test

edbeeching committed:
Set done and info to _ to fix failing test

patrickvonplaten commented on tests/test_modeling_decision_transformer.py:
```diff
@@ -0,0 +1,251 @@
+        NUM_STEPS = 2  # number of steps of autoregressive prediction we will perform
+        TARGET_RETURN = 10  # defined by the RL environment, may be normalized
+        config = DecisionTransformerConfig()
+        model = DecisionTransformerModel(config).to(torch_device)
```
Do you think we could use a pretrained model here instead? To be sure that the weights are not randomly initialized .

patrickvonplaten commented on tests/test_modeling_decision_transformer.py:
```diff
@@ -0,0 +1,251 @@
+
+            self.assertEqual(action_pred.shape, actions.shape)
+
+            state, reward, _, _ = (  # env.step(action)
```
```suggestion
            torch.manual_seed(0)
            state, reward, _, _ = (  # env.step(action)
```

to make it deterministic

patrickvonplaten commented on tests/test_modeling_decision_transformer.py:
```diff
@@ -0,0 +1,251 @@
+                {},
+            )
+
+            actions[-1] = action_pred[0, -1]
```
think here we can really check that `actions[-1]` for its numerical values (kind of like we're doing it here: https://github.com/huggingface/transformers/blob/93d3fd86459eb27bc584da29a3d542817a395bca/tests/bert/test_modeling_bert.py#L593 for BERT's hidden states) if we make everything before deterministic.

edbeeching committed:
Updated integration test to be deterministic and check expected outputs

edbeeching committed:
Updated integration test to be deterministic and check expected outputs

sgugger commented on docs/source/index.mdx:
```diff
@@ -183,6 +183,7 @@ Flax), PyTorch, and/or TensorFlow.
 |            CTRL             |       ✅       |       ❌       |       ✅        |         ✅         |      ❌      |
 |           DeBERTa           |       ✅       |       ✅       |       ✅        |         ✅         |      ❌      |
 |         DeBERTa-v2          |       ✅       |       ❌       |       ✅        |         ✅         |      ❌      |
+|     DecisionTransformer     |       ❌       |       ❌       |       ✅        |         ❌         |      ❌      |
```
This model needs to also be added in the main README.

sgugger commented on docs/source/model_doc/decision_transformer.mdx:
```diff
@@ -0,0 +1,48 @@
+specific language governing permissions and limitations under the License.
+-->
+
+# DecisionTransformer
```
```suggestion
# Decision Transformer
```
(This applies to all instances in this file.)

sgugger commented on docs/source/model_doc/decision_transformer.mdx:
```diff
@@ -0,0 +1,48 @@
+
+## Overview
+
+The DecisionTransformer model was proposed in [Decision Transformer: Reinforcement Learning via Sequence Modeling
+](https://arxiv.org/abs/2106.01345)  
```
```suggestion
The DecisionTransformer model was proposed in [Decision Transformer: Reinforcement Learning via Sequence Modeling](https://arxiv.org/abs/2106.01345)  
```

sgugger commented on src/transformers/__init__.py:
```diff
@@ -650,6 +651,7 @@
     _import_structure["modeling_utils"] = ["Conv1D", "PreTrainedModel", "apply_chunking_to_forward", "prune_layer"]
 
     # PyTorch models structure
+
```
```suggestion
```

sgugger commented on src/transformers/__init__.py:
```diff
@@ -2781,6 +2794,7 @@
         from .utils.dummy_pytorch_quantization_and_torch_objects import *
 
     if is_torch_available():
+
```
```suggestion
```

sgugger commented on src/transformers/modeling_outputs.py:
```diff
@@ -245,6 +245,41 @@ class BaseModelOutputWithPastAndCrossAttentions(ModelOutput):
+@dataclass
+class DecisionTransformerOutput(ModelOutput):
+    """
+    Base class for model's outputs that also contains a pooling of the last hidden states. TODO
```
```suggestion
    Base class for model's outputs that also contains a pooling of the last hidden states.
```
That is left to do here?

edbeeching commented:
Thanks for the comments, the model is now loaded from a pretrained checkpoints, I have set the seed and have added a comparison between the output and the expected output. Let me know of any other comments or additions.

sgugger commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -211,6 +213,7 @@
 MODEL_NAMES_MAPPING = OrderedDict(
     [
         # Add full (and cased) model names here
+        ("decision_transformer", "DecisionTransformer"),
```
```suggestion
        ("decision_transformer", "Decision Transformer"),
```

sgugger commented on src/transformers/models/decision_transformer/configuration_decision_transformer.py:
```diff
@@ -0,0 +1,205 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" DecisionTransformer model configuration"""
```
```suggestion
""" Decision Transformer model configuration"""
```

sgugger commented on src/transformers/models/decision_transformer/configuration_decision_transformer.py:
```diff
@@ -0,0 +1,205 @@
+logger = logging.get_logger(__name__)
+
+DECISION_TRANSFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "decision_transformer": "https://huggingface.co/decision_transformer/resolve/main/config.json",
```
That url and checkpoint name can't work as they are missing an organization.

edbeeching replied:
Is this not the same for other models in the transformers library, such as [gpt2](https://github.com/huggingface/transformers/blob/master/src/transformers/models/gpt2/configuration_gpt2.py#L30) ?

patrickvonplaten replied:
```suggestion
    "decision_transformer": "https://huggingface.co/edbeeching/decision-transformer-gym-hopper-medium/resolve/main/config.json",
```

sgugger replied:
This needs to be addressed before merging. Patrick's suggestion is still missing the namespace in the key, which should be `edbeeching/decision-transformer-gym-hopper-medium` to match the model.

sgugger commented on src/transformers/models/decision_transformer/configuration_decision_transformer.py:
```diff
@@ -0,0 +1,205 @@
+class DecisionTransformerConfig(PretrainedConfig):
+    """
+    This is the configuration class to store the configuration of a [`DecisionTransformerModel`]. It is used to
+    instantiate a DecisionTransformer model according to the specified arguments, defining the model architecture.
```
```suggestion
    instantiate a Decision Transformer model according to the specified arguments, defining the model architecture.
```

sgugger commented on src/transformers/models/decision_transformer/configuration_decision_transformer.py:
```diff
@@ -0,0 +1,205 @@
+            Number of hidden layers in the Transformer encoder.
+        n_head (`int`, *optional*, defaults to 12):
+            Number of attention heads for each attention layer in the Transformer encoder.
+        n_inner (`int`, *optional*, defaults to None):
```
```suggestion
        n_inner (`int`, *optional*):
```

sgugger commented on src/transformers/models/decision_transformer/configuration_decision_transformer.py:
```diff
@@ -0,0 +1,205 @@
+        n_head (`int`, *optional*, defaults to 12):
+            Number of attention heads for each attention layer in the Transformer encoder.
+        n_inner (`int`, *optional*, defaults to None):
+            Dimensionality of the inner feed-forward layers. `None` will set it to 4 times n_embd
```
```suggestion
            Dimensionality of the inner feed-forward layers. If unset, will default to 4 times `n_embd`.
```

sgugger commented on src/transformers/models/decision_transformer/configuration_decision_transformer.py:
```diff
@@ -0,0 +1,205 @@
+                - `"mean"`: Take the mean of all tokens hidden states.
+                - `"cls_index"`: Supply a Tensor of classification token position (like GPT/GPT-2).
+                - `"attn"`: Not implemented now, use multi-head attention.
+        summary_use_proj (`bool`, *optional*, defaults to `True`):
+            Argument used when doing sequence summary, used in the models [`GPT2DoubleHeadsModel`] and
+            [`TFGPT2DoubleHeadsModel`].
+
+            Whether or not to add a projection after the vector extraction.
+        summary_activation (`str`, *optional*):
+            Argument used when doing sequence summary. Used in for the multiple choice head in
+            [`GPT2DoubleHeadsModel`].
+
+            Pass `"tanh"` for a tanh activation to the output, any other value will result in no activation.
+        summary_proj_to_labels (`bool`, *optional*, defaults to `True`):
+            Argument used when doing sequence summary, used in the models [`GPT2DoubleHeadsModel`] and
+            [`TFGPT2DoubleHeadsModel`].
+
+            Whether the projection outputs should have `config.num_labels` or `config.hidden_size` classes.
+        summary_first_dropout (`float`, *optional*, defaults to 0.1):
+            Argument used when doing sequence summary, used in the models [`GPT2DoubleHeadsModel`] and
+            [`TFGPT2DoubleHeadsModel`].
+
+            The dropout ratio to be used after the projection and activation.
```
Those are used for models with head in GPT-2, it doesn't look like you are using those here, so would remove.

sgugger commented on src/transformers/models/decision_transformer/modeling_decision_transformer.py:
```diff
@@ -0,0 +1,224 @@
+from ...modeling_outputs import DecisionTransformerOutput
+from ...modeling_utils import PreTrainedModel
+from ...utils import logging
+from ..gpt2 import GPT2Model
```
This breaks the core philosophy of Transformers that each model file should be fully independent from the others, so we never import other components in a modeling file. Instead we copy the blocks we need and add `# Copied from` statements.

patrickvonplaten replied:
Agree here with @sgugger I think. Sorry @edbeeching I might have been a bit misleading in my messages on Slack. 

To be more precise, I think we should never import a specfic model like `GPT2Model` into another modeling file. I think there are two possibilities here:

1. Decision Transformers only works with GPT2 (all it's checkpoints are GPT2-like checkpoints). In this case I 100% agree with @sgugger and we should just copy-paste the whole GPT2Model architecture under the name `DecisionTransformerModel` into this file and add a nice `# Copied from ...` line to ensure the code is kept up to date.
2. If we have checkpoints for GPT2, GPT-J and GPT-Neo, then I can be convinced into importing the `AutoModelForCausalLM` class here and make the model a composition model.

However given that the model is just a thin wrapper above GPT2 (e.g. essentially just uses GPT2 as the encoder), I also prefer 1. here.


Sorry, @edbeeching I might have been misleading on Slack (first time I'm doing a full review on the PR)

edbeeching replied:
No problem. Do you have an example where this has been done before? So I do not have to write/copy/adjust too much boilerplate. Should I call it DecisionTransformerGPT2Model ? I guess I need to copy all the tests as well?

sgugger commented on src/transformers/models/decision_transformer/modeling_decision_transformer.py:
```diff
@@ -0,0 +1,224 @@
+    """
+
+    config_class = DecisionTransformerConfig
+    # load_tf_weights = load_tf_weights_in_decision_transformer
```
Should be cleaned up.

sgugger commented on src/transformers/models/decision_transformer/modeling_decision_transformer.py:
```diff
@@ -0,0 +1,224 @@
+
+_CHECKPOINT_FOR_DOC = "decision_transformer"
+_CONFIG_FOR_DOC = "DecisionTransformerConfig"
+# _TOKENIZER_FOR_DOC = "DecisionTransformerTokenizer"
```
Should be cleaned up.

sgugger commented on src/transformers/models/decision_transformer/modeling_decision_transformer.py:
```diff
@@ -0,0 +1,224 @@
+"""
+
+
+@add_start_docstrings(
+    "The DecisionTransformer Model",
+    DECISION_TRANSFORMER_START_DOCSTRING,
+)
```
```suggestion
@add_start_docstrings("The Decision Transformer Model", DECISION_TRANSFORMER_START_DOCSTRING)
```

sgugger commented on tests/test_modeling_decision_transformer.py:
```diff
@@ -0,0 +1,254 @@
+        self.parent.assertEqual(result.action_preds.shape, actions.shape)
+        self.parent.assertEqual(result.return_preds.shape, returns_to_go.shape)
+        self.parent.assertEqual(
+            result.last_hidden_state.shape,
+            (
+                self.batch_size,
+                self.seq_length * 3,
+                self.hidden_size,
+            ),  # seq length *3 as there are 3 modelities: states, returns and actions
```
Put the comment elsewhere so this can all fit in one line.

sgugger commented on tests/test_modeling_decision_transformer.py:
```diff
@@ -0,0 +1,254 @@
+
+
+@require_torch
+class DecisionTransformerModelTest(  # ModelTesterMixin is removed as this model as the input / outputs of this model do not conform to a typical NLP model.
```
Except it's right below, so remove the comment?

sgugger reviewed (commented):
Thanks for the model addition. I've left a few nits here and there (in particular we should respect the name selected by the authors, which is Decision Transformers with a space).

My main concern is around the modeling file code. The core principle of the library is that each modeling file should be independent from the others, and we have designed mechanisms to make sure copy-pasted code stays in sync across files. We make an exception to that rule for models that works for several blocks (like encoder-decoder models, or models that can accept several architectures as a "backbone") but it doesn't seem to be the case here, so I'm curious why you're not pushing to include the code of GPT-2 @patrickvonplaten @LysandreJik

patrickvonplaten commented on tests/test_modeling_decision_transformer.py:
```diff
@@ -0,0 +1,254 @@
+        torch.manual_seed(0)
+        state = torch.randn(1, 1, config.state_dim).to(device=torch_device, dtype=torch.float32)  # env.reset()
+
+        expected_outputs = torch.tensor([[0.2384, -0.2955, 0.8741], [0.6765, -0.0793, -0.1298]], device=torch_device)
```
Very nice!

edbeeching committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edbeeching committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edbeeching committed:
Removed unnecessary config options

edbeeching committed:
Removed unnecessary config options

edbeeching committed:
Cleaned up commented code and old comments.

edbeeching committed:
Cleaned up commented code and old comments.

edbeeching committed:
Cleaned up commented code.

edbeeching committed:
Cleaned up commented code.

edbeeching committed:
Changed DecisionTransformer to Decision Transformer

edbeeching committed:
Changed DecisionTransformer to Decision Transformer

edbeeching committed:
Added Decision Transformer to the main README file

edbeeching committed:
Added Decision Transformer to the main README file

edbeeching committed:
Added copy of GTP2 called DecisionTranformerGPT2Model

edbeeching committed:
Added copy of GTP2 called DecisionTranformerGPT2Model

edbeeching committed:
isorted imports

edbeeching committed:
isorted imports

edbeeching committed:
isorted imports

edbeeching committed:
isorted imports

edbeeching committed:
Added model to non-English README files

edbeeching committed:
Added model to non-English README files

edbeeching committed:
Ran make fix-copies and corrected some cases.

edbeeching committed:
Ran make fix-copies and corrected some cases.

edbeeching committed:
Updated index file to include Decision Transformer

edbeeching committed:
Updated index file to include Decision Transformer

edbeeching committed:
Added gpt2 model as copy inside the Decision Transformer model file

edbeeching committed:
Added gpt2 model as copy inside the Decision Transformer model file

edbeeching committed:
Added the unit test file to the list of TEST_FILES_WITH_NO_COMMON_TESTS

edbeeching committed:
Added the unit test file to the list of TEST_FILES_WITH_NO_COMMON_TESTS

patrickvonplaten commented on MUJOCO_LOG.TXT:
```diff
@@ -0,0 +1,3 @@
+Tue Mar 15 11:38:16 2022
```
Think this file we should delete no?

patrickvonplaten replied:
This still needs to be deleted no?

patrickvonplaten commented on zeb_dt_convert_checkpoints.py:
```diff
@@ -0,0 +1,84 @@
+import gym
+from transformers import DecisionTransformerModel, DecisionTransformerConfig
+
+# original_checkpoint = torch.load("dt_hopper_old.pth")
```
should be delete this? Also let's maybe put this file under `src/transformers/models/decision_transformer` no?

patrickvonplaten commented on zeb_dt_test.py:
```diff
@@ -0,0 +1,98 @@
+import random
```
should we put this file as well under `src/transformers/models/decision_transformers`?

patrickvonplaten reviewed (commented):
In general this looks good to me! Thanks a lot for adding this! 

A couple of final changes we should do IMO:
- delete all those `decision-transformer-gym` folder (I think they were added accidently no?)
- Move the model checkpoints to a "UC Berkely" org - would that be ok for you @edbeeching ? 

Apart from this, the PR looks ready for merge to me!

edbeeching committed:
Deleted redundant checkpoint files (I don't know how these got committed)

edbeeching committed:
Deleted redundant checkpoint files (I don't know how these got committed)

edbeeching committed:
Removed testing files. (These should have never been committed)

edbeeching committed:
Removed testing files. (These should have never been committed)

patrickvonplaten commented:
> checkpoint

I see! Yeah since you've trained them, then I think it's totally fine to just publish them under your name on the Hub! So I think the PR is good to go for me :-)

patrickvonplaten commented:
Think we still got some `decision-transformer-gym-halfcheetah-expert` folders though ;-)

edbeeching committed:
Removed accidentally committed files

edbeeching committed:
Removed accidentally committed files

patrickvonplaten reviewed (approved):
Looks good to me! @sgugger could you maybe take a final look? :-)

sgugger commented:
There are five conflicts to resolve, could you take care of this before we do a final review @edbeeching ?

edbeeching committed:
Merge branch 'master' into decision_transformer_gym

edbeeching commented:
Ok I went through the conflicts.
One final question, it appears that the test files have been reorganized since I first forked the transformers repo. They now appear to be nested each in their own subdirectory. Should add a directory for my test to reflect this change in organization?

edbeeching committed:
Moved the Decision Transformer test to its own directory

edbeeching committed:
Moved the Decision Transformer test to its own directory

Tegzes committed:
Add type hints for Pegasus (#16324)

Commit:
Funnel type hints (#16323)

* add pt funnel type hints

* add tf funnel type hints

Tegzes committed:
Add type hints for ProphetNet PyTorch (#16272)

sgugger commented on README.md:
```diff
@@ -252,6 +252,7 @@ Current number of checkpoints: ![](https://img.shields.io/endpoint?url=https://h
 1. **[Data2Vec](https://huggingface.co/docs/transformers/master/model_doc/data2vec)** (from Facebook) released with the paper [Data2Vec:  A General Framework for Self-supervised Learning in Speech, Vision and Language](https://arxiv.org/abs/2202.03555) by Alexei Baevski, Wei-Ning Hsu, Qiantong Xu, Arun Babu, Jiatao Gu, Michael Auli.
 1. **[DeBERTa](https://huggingface.co/docs/transformers/model_doc/deberta)** (from Microsoft) released with the paper [DeBERTa: Decoding-enhanced BERT with Disentangled Attention](https://arxiv.org/abs/2006.03654) by Pengcheng He, Xiaodong Liu, Jianfeng Gao, Weizhu Chen.
 1. **[DeBERTa-v2](https://huggingface.co/docs/transformers/model_doc/deberta-v2)** (from Microsoft) released with the paper [DeBERTa: Decoding-enhanced BERT with Disentangled Attention](https://arxiv.org/abs/2006.03654) by Pengcheng He, Xiaodong Liu, Jianfeng Gao, Weizhu Chen.
+1. **[Decision Transformer](https://huggingface.co/docs/transformers/model_doc/DecisionTransformer)** (from Berkeley/Facebook/Google) released with the paper [Decision Transformer: Reinforcement Learning via Sequence Modeling](https://arxiv.org/abs/2106.01345) by Lili Chen, Kevin Lu, Aravind Rajeswaran, Kimin Lee, Aditya Grover, Michael Laskin, Pieter Abbeel, Aravind Srinivas, Igor Mordatch.
```
```suggestion
1. **[Decision Transformer](https://huggingface.co/docs/transformers/master/model_doc/decision_transformer)** (from Berkeley/Facebook/Google) released with the paper [Decision Transformer: Reinforcement Learning via Sequence Modeling](https://arxiv.org/abs/2106.01345) by Lili Chen, Kevin Lu, Aravind Rajeswaran, Kimin Lee, Aditya Grover, Michael Laskin, Pieter Abbeel, Aravind Srinivas, Igor Mordatch.
```

The model won't be in the stable doc until next release. Plus the page will have the same name as the MDX, namely `decision_transformer`.

sgugger commented on examples/pytorch/decision_transformer/run_decision_transformer.py:
```diff
@@ -0,0 +1,182 @@
+import numpy as np
```
We don't usually put examples that work for just one model in the examples folder, but in a research project. I think this one should be moved there.

sgugger commented on examples/pytorch/decision_transformer/run_decision_transformer.py:
```diff
@@ -0,0 +1,182 @@
+state_mean = torch.from_numpy(state_mean).to(device=device)
+state_std = torch.from_numpy(state_std).to(device=device)
+
+# Create the decision transformer model
+# config = DecisionTransformerConfig(state_dim=state_dim, act_dim=act_dim, max_ep_len=1000, max_length=20)
+# model = DecisionTransformerModel(config)
+model = DecisionTransformerModel.from_pretrained("edbeeching/decision-transformer-gym-hopper-medium")
+model = model.to(device)
+model.eval()
+# Load the pretrained weights from a checkpoint
+# save a checkpoint of this model to compare keys
+# torch.save(model.state_dict(), "dt_hopper.pth")
+# state_dict = torch.load("checkpoints/hopper_medium.pth")
+# model.load_state_dict(state_dict)
+# Interact with the environment
```
Lots of comments to clean up (or if we leave them, we need more explanation around them).

sgugger commented on src/transformers/modeling_outputs.py:
```diff
@@ -245,6 +245,41 @@ class BaseModelOutputWithPastAndCrossAttentions(ModelOutput):
 
 
+@dataclass
+class DecisionTransformerOutput(ModelOutput):
```
This is specific to Decision Transformers, so it should really go in the `modeling_decision_transformers` file.

sgugger commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -28,6 +28,8 @@
     [
         # Base model mapping
+        ("decision_transformer", "DecisionTransformerModel"),
+        ("decision_transformer_gpt2", "DecisionTransformerGPT2Model"),
```
```suggestion
```
This model type does not exist, so this shouldn't be here. It's okay if `DecisionTransformerGPT2Model` is not in an auto mapping.

sgugger commented on src/transformers/models/decision_transformer/configuration_decision_transformer.py:
```diff
@@ -0,0 +1,185 @@
+            Whether to use a tanh activation on action prediction
+        vocab_size (`int`, *optional*, defaults to 50257):
+            Vocabulary size of the GPT-2 model. Defines the number of different tokens that can be represented by the
+            `inputs_ids` passed when calling [`GPT2Model`] or [`TFGPT2Model`].
```
```suggestion
            `inputs_ids` passed when calling [`DecisionTransformerModel`].
```

sgugger commented on src/transformers/models/decision_transformer/configuration_decision_transformer.py:
```diff
@@ -0,0 +1,185 @@
+            The epsilon to use in the layer normalization layers.
+        initializer_range (`float`, *optional*, defaults to 0.02):
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        summary_type (`string`, *optional*, defaults to `"cls_index"`):
+            Argument used when doing sequence summary, used in the models [`GPT2DoubleHeadsModel`] and
+            [`TFGPT2DoubleHeadsModel`].
+
+            Has to be one of the following options:
+
+                - `"last"`: Take the last token hidden state (like XLNet).
+                - `"first"`: Take the first token hidden state (like BERT).
+                - `"mean"`: Take the mean of all tokens hidden states.
+                - `"cls_index"`: Supply a Tensor of classification token position (like GPT/GPT-2).
+                - `"attn"`: Not implemented now, use multi-head attention.
```
This is only used by head models, so should be removed from this config.

sgugger reviewed (commented):
Thanks for iterating on the PR!

There are still a few things to clean up, left comments. Also, don't forget to run `make style` on your branch to make the quality tests pass, and your new test subfolder is missing an `init`, I believe, which makes most tests fail.

NielsRogge committed:
[GLPN] Improve docs (#16331)

* Add link to notebook

* Add link

* Fix bug

Co-authored-by: Niels Rogge <nielsrogge@Nielss-MacBook-Pro.local>

clefourrier committed:
Added type hints for Pytorch Marian calls (#16200)

* Added type hinting for forward functions in pytorch marian

* typo correction

* Removed type hints on functions from BART per Suraj Patil request

* fix import pb

* fix typo

* corrected tuple call

* ran black

* after fix-copies
Some optional tags on primitives were removed, past_key_values in MarianForCausalLM changed from Tuple of Tuple to List

* Fixing copies to roformer and pegasus

Co-authored-by: Clementine Fourrier <cfourrie@inria.fr>
Co-authored-by: matt <rocketknight1@gmail.com>

edbeeching committed:
Moved DecisionTransformOutput to modeling_decision_transformer

edbeeching committed:
Moved DecisionTransformOutput to modeling_decision_transformer

edbeeching committed:
Moved the example usage to research project and cleaned comments

edbeeching committed:
Moved the example usage to research project and cleaned comments

edbeeching committed:
Made tests ignore the copy of gpt2 in Decision Transformer

edbeeching committed:
Made tests ignore the copy of gpt2 in Decision Transformer

edbeeching committed:
Added module output to modelling decision transformer

edbeeching committed:
Added module output to modelling decision transformer

edbeeching committed:
removed copied gpt2 model from list of transformers models

edbeeching committed:
removed copied gpt2 model from list of transformers models

edbeeching committed:
Updated tests and created __init__ file for new test location

edbeeching committed:
Updated tests and created __init__ file for new test location

edbeeching committed:
Update README.md

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edbeeching committed:
Update README.md

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edbeeching committed:
Update src/transformers/models/decision_transformer/configuration_decision_transformer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edbeeching committed:
Update src/transformers/models/decision_transformer/configuration_decision_transformer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edbeeching committed:
Removed unneeded summary type from config file

edbeeching committed:
Removed unneeded summary type from config file

edbeeching committed:
Fixed copies

edbeeching committed:
Fixed copies

edbeeching committed:
Updated pretrained config map to refer to hopper-medium checkpoint

edbeeching committed:
Updated pretrained config map to refer to hopper-medium checkpoint

FrancescoSaverioZuppichini committed:
done (#16340)

edbeeching committed:
Added Decision transformer to model docs

edbeeching committed:
Added Decision transformer to model docs

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_15845). All of your documentation changes will be reflected on that endpoint.

sgugger commented on src/transformers/models/decision_transformer/modeling_decision_transformer.py:
```diff
@@ -0,0 +1,917 @@
+
+logger = logging.get_logger(__name__)
+
+_CHECKPOINT_FOR_DOC = "decision_transformer"
```
```suggestion
_CHECKPOINT_FOR_DOC = "edbeeching/decision-transformer-gym-hopper-medium"
```

sgugger commented on src/transformers/models/decision_transformer/modeling_decision_transformer.py:
```diff
@@ -0,0 +1,917 @@
+_CONFIG_FOR_DOC = "DecisionTransformerConfig"
+
+DECISION_TRANSFORMER_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "decision_transformer",
```
```suggestion
    "edbeeching/decision-transformer-gym-hopper-medium",
```

sgugger commented on src/transformers/models/decision_transformer/configuration_decision_transformer.py:
```diff
@@ -0,0 +1,174 @@
+logger = logging.get_logger(__name__)
+
+DECISION_TRANSFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "decision-transformer-gym-hopper-medium": "https://huggingface.co/edbeeching/decision-transformer-gym-hopper-medium/resolve/main/config.json",
```
```suggestion
    "edbeeching/decision-transformer-gym-hopper-medium": "https://huggingface.co/edbeeching/decision-transformer-gym-hopper-medium/resolve/main/config.json",
```

sgugger reviewed (approved):
Thanks for all the changes! Just a last few nits on the checkpoint names, otherwise good to merge!

edbeeching committed:
Update src/transformers/models/decision_transformer/modeling_decision_transformer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edbeeching committed:
Update src/transformers/models/decision_transformer/modeling_decision_transformer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edbeeching committed:
Update src/transformers/models/decision_transformer/modeling_decision_transformer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edbeeching committed:
Update src/transformers/models/decision_transformer/modeling_decision_transformer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edbeeching committed:
Update src/transformers/models/decision_transformer/configuration_decision_transformer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edbeeching committed:
Update src/transformers/models/decision_transformer/configuration_decision_transformer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jacobdineen committed:
Add type annotations for Rembert/Splinter and copies (#16338)

* undo black autoformat

* minor fix to rembert forward with default

* make fix-copies, make quality

* Adding types to template model

* Removing List from the template types

* Remove `Optional` from a couple of types that don't accept `None`

Co-authored-by: matt <rocketknight1@gmail.com>

patrickvonplaten committed:
[Bug template] Shift responsibilities for long-range (#16344)

osanseviero committed:
Fix code repetition in serialization guide (#16346)

stevhliu committed:
Adopt framework-specific blocks for content (#16342)

* ✨ refactor code samples with framework-specific blocks

* ✨ update training.mdx

* 🖍 apply feedback

LysandreJik committed:
Updates the default branch from master to main (#16326)

* Updates the default branch from master to main

* Links from `master` to `main`

* Typo

* Update examples/flax/README.md

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edbeeching committed:
Updated model with custom docstring example

edbeeching committed:
Updated model with custom docstring example

edbeeching committed:
Merge remote-tracking branch 'origin/decision_transformer_gym' into decision_transformer_gym

edbeeching committed:
Updated copies, config auto, and readme files.
